spec/03-plugins/23-aws-lambda/01-access_spec.lua

```diff
@@ -527,5 +553,93 @@ for _, strategy in helpers.each_strategy() do
       assert.equal(65, tonumber(res.headers["Content-Length"]))
     end)
 
+    it("returns the proper status code when setting custom response body from Lambda", function()
+      -- the lambda function must return a string
+      -- for the custom response "body" property
+      local body = cjson.encode({
+        key1 = "some_value_post1",
+        key2 = "some_value_post2",
+        key3 = "some_value_post3"
+      })
+      local res = assert(proxy_client:send {
+        method  = "POST",
+        path    = "/post",
+        headers = {
+          ["Host"]         = "lambda11.com",
+          ["Content-Type"] = "application/json"
+        },
+        body = {
+          statusCode = 201,
+          body = body,
+          headers = {
+            ["X-Custom-Header"] = "Hello world!"
+          }
+        }
+      })
+      local res_body = assert.res_status(201, res)
+      assert.equal(79, tonumber(res.headers["Content-Length"]))
+      assert.equal("Hello world!", res.headers["X-Custom-Header"])
+      assert.equal(body, res_body)
+    end)
+
+    it("returns proper status code with no body or headers", function()
+      local res = assert(proxy_client:send {
+        method  = "POST",
+        path    = "/post",
+        headers = {
+          ["Host"]         = "lambda11.com",
+          ["Content-Type"] = "application/json"
+        },
+        body = {
+          statusCode = 201,
+        }
+      })
+      local body = assert.res_status(201, res)
+      assert.equal(0, tonumber(res.headers["Content-Length"]))
+      assert.equal(nil, res.headers["X-Custom-Header"])
+      assert.equal("", body)
+    end)
+
+    it("override duplicate headers with value from the custom response", function()
+      -- the default "x-amzn-RequestId" returned is "foo"
+      -- let's check it is overriden with a custom value
+      local headers = {
+        ["x-amzn-RequestId"] = "bar"
+      }
+      local res = assert(proxy_client:send {
+        method  = "POST",
+        path    = "/post",
+        headers = {
+          ["Host"]         = "lambda11.com",
+          ["Content-Type"] = "application/json"
+        },
+        body = {
+          statusCode = 201,
+          headers = headers,
+        }
+      })
+
+      assert.res_status(201, res)
+      assert.equal("bar", res.headers["x-amzn-RequestId"])
+    end)
+
+    it("errors with 502 with badly formed response from Lambda", function()
+      local res = assert(proxy_client:send {
+        method  = "POST",
+        path    = "/post",
+        headers = {
+          ["Host"]         = "lambda11.com",
+          ["Content-Type"] = "application/json"
+        },
+        body = {
+          statusCode = 201,
+          body = 1234,
+        }
```

Comment:
We might be missing a test case when the actual response from the Lambda is _not_ a JSON payload, as pointed by my earlier comment). Here, only the value of `body` is not JSON.